### PR TITLE
PROGRAMMES-5758 footer watchlive

### PR DESCRIPTION
**Pull request by mattgreenham**
https://jira.dev.bbc.co.uk/browse/PROGRAMMES-5758

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: app/Resources/js/lazyload.js [hunk 1]
```diff
@@ -13,16 +29,18 @@ define(['jquery-1.9'], function ($) {
...
+            data_delay_lazyload : 'data-lazyload-delay'
         },
         setOptions: function (options) {
             this.options = $.extend({}, this.options, options);
         },
         init: function () {
             var _this = this;
 
-            this.loadContents();
+            _this.loadContents();
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: This shouldn't need changing, should it?

File: app/Resources/js/lazyload.js [hunk 2]
```diff
@@ -34,18 +52,71 @@ define(['jquery-1.9'], function ($) {
...
+            var window_top = $window.scrollTop();
+            var offset = $element.offset();
+            var left = offset.left;
+            var top = offset.top;
+
+            // load just before in viewport
+            var lazy_top_offset = 500;
+            var lazy_left_offset = 500;
+
+            if (top + $element.height() >= window_top &&
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: Can the logic in this `if` be extracted to a method to make it more readable?

File: app/Resources/js/lazyload.js [hunk 3]
```diff
@@ -34,18 +52,71 @@ define(['jquery-1.9'], function ($) {
...
+                top - lazy_top_offset <= window_top + $window.height() &&
+                left + $element.width() >= window_left &&
+                left - lazy_left_offset <= window_left + $window.width()) {
+                return true;
+            } else {
+                return false;
+            }
+        },
+        addOnScrollHandler : function (context) {
+            var _this = this;
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: Why do we keep doing this instead of `.bind(this)`?

File: app/Resources/js/lazyload.js [hunk 4]
```diff
@@ -34,18 +52,71 @@ define(['jquery-1.9'], function ($) {
...
+
+            var check_in_view = debounce(function () {
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: Why is this function assigned to a variable? Why is it not just a named method?

File: app/Resources/js/lazyload.js (hunk 2 above)
Comment (mattgreenham) [reply app/Resources/js/lazyload.js]: I don't think so, that's what this entire method does?

Comment (votemike) [reply app/Resources/js/lazyload.js]: Extract each line to a variable instead.  
Then `return a && b && c && d;`

File: app/Resources/js/lazyload.js [hunk 5]
```diff
@@ -4,6 +4,22 @@ define(['jquery-1.9'], function ($) {
         this.setOptions(options);
     };
 
+    var toAppear = [];
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: What is this an array of? I can't figure it out.

File: app/Resources/js/lazyload.js [hunk 6]
```diff
@@ -4,6 +4,22 @@ define(['jquery-1.9'], function ($) {
...
+
+    function debounce(func, wait) {
+        var timeout;
+        return function() {
+            var context = this, args = arguments;
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: Does `this` need to be assigned to `context`? Can't you just pass `this` straight in to `func.apply()`?

File: app/Resources/js/lazyload.js [hunk 7]
```diff
@@ -34,18 +52,68 @@ define(['jquery-1.9'], function ($) {
...
+                    if (!this.isDelayed(content) || this.isInViewport(content)) {
+                        this.loadAjaxContent(content);
+                    }
+                    else {
+                        toAppear.push(content);
+                    }
                 }
             }
+
+            this.addOnScrollOrResizeHandler(context);
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: Could you explain why this is here. And what `context` is?

File: app/Resources/js/lazyload.js [hunk 8]
```diff
@@ -34,18 +52,68 @@ define(['jquery-1.9'], function ($) {
...
+            var right_of_lhs = (element_left + $element.width() >= window_left);
+            var left_of_rhs = (element_left - lazy_left_offset <= window_left + $window.width());
+
+            return below_window_top && above_window_bottom && right_of_lhs && left_of_rhs;
+        },
+        checkInView : function (context) {
+            return debounce(function () {
+                if (toAppear.length > 0) {
+                    var content;
+                    for (var i = 0; (content = toAppear[i]) != null; i++) {
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: Is this not better-off as a `.each()`?

File: app/Resources/js/lazyload.js (hunk 5 above)
Comment (mattgreenham) [reply app/Resources/js/lazyload.js]: Lazy elements which haven't yet been loaded i.e. not yet in view. On page load we find all the 'lazy-module' elements on the page so on scroll we only need to check if elements in this array are in view.

File: app/Resources/js/lazyload.js (hunk 6 above)
Comment (mattgreenham) [reply app/Resources/js/lazyload.js]: No. Yes. Changed.

File: app/Resources/js/dsamen-bootstrap.js [hunk 9]
```diff
@@ -10,6 +10,9 @@ define(['eqjs', 'respimg', 'lazysizes'], function(EQ) {
             window.addEventListener('lazyload-complete',function(e){
                 EQ.query(document.querySelectorAll(selector));
             },false);
+
+            var lazyload = new Lazyload();
```
Comment (tatitati) [inline app/Resources/js/dsamen-bootstrap.js]: I suspect that is not necessary to create a new lazyload object. You are passing already an object in there. Just try (maybe I'm wrong)

File: app/Resources/js/lazyload.js (hunk 5 above)
Comment (votemike) [reply app/Resources/js/lazyload.js]: So.... a bunch or jQuery objects? A bunch of HTML tags? What?

Comment (mattgreenham) [reply app/Resources/js/lazyload.js]: jQuery objects

File: app/Resources/js/dsamen-bootstrap.js (hunk 9 above)
Comment (mattgreenham) [reply app/Resources/js/dsamen-bootstrap.js]: Doesn't work without it

Comment (tatitati) [reply app/Resources/js/dsamen-bootstrap.js]: Interesting. Thanks Matt. Sorry for my "experimental feedback" jeje

File: app/Resources/js/lazyload.js (hunk 5 above)
Comment (votemike) [reply app/Resources/js/lazyload.js]: Add a comment to this, I still get confused as to what it is.

File: app/Resources/js/lazyload.js [hunk 10]
```diff
@@ -34,18 +51,66 @@ define(['jquery-1.9'], function ($) {
...
+
+            return below_window_top && above_window_bottom && right_of_lhs && left_of_rhs;
+        },
+        checkInView : function () {
+            return debounce(function () {
+                var _this = this;
+                toAppear.forEach(function (element, i) {
+                    if (_this.shouldBeRenderedNow(element)) {
+                        _this.loadAjaxContent(element);
+                        toAppear.splice(i, 1);
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: `i` looks to similar to `1`. Rename it. Maybe to `key`?

File: app/Resources/js/lazyload.js [hunk 11]
```diff
@@ -34,18 +51,66 @@ define(['jquery-1.9'], function ($) {
 
             return (lazy_contents.length) ? lazy_contents : false;
         },
+        isOverThreshold : function (lazyloaded_object) {
+            var threshold = lazyloaded_object.getAttribute(this.options.data_threshold);
```
Comment (votemike) [inline app/Resources/js/lazyload.js]: Can you rename `threshold` to `element_threshold`?

Comment (votemike) [reply app/Resources/js/lazyload.js]: Or maybe even `minimum_render_threshold` or something similar.

**Comment by tatitati** [comment]
COOOOOOOOOOOOOOOOL

**Pull request merged**